pkg/injector/bootstrap.go

```diff
@@ -100,3 +105,103 @@ func (wh *mutatingWebhook) marshalAndSaveBootstrap(name, namespace string, confi
 	log.Debug().Msgf("Creating bootstrap config for Envoy: name=%s, namespace=%s", name, namespace)
 	return wh.kubeClient.CoreV1().Secrets(namespace).Create(context.Background(), secret, metav1.CreateOptions{})
 }
+
+// NewBootstrapSecretRotator returns a new bootstrap secret rotator.
+func NewBootstrapSecretRotator(ctx context.Context, kubeClient kubernetes.Interface, informerCollection *informers.InformerCollection, certManager *certificate.Manager, checkInterval time.Duration) *BootstrapSecretRotator {
+	return &BootstrapSecretRotator{
+		context:            ctx,
+		kubeClient:         kubeClient,
+		informerCollection: informerCollection,
+		certManager:        certManager,
+		checkInterval:      checkInterval,
+	}
+}
+
+// listBootstrapSecrets returns the bootstrap secrets stored in the informerCollection's store.
+// this function is used to get the namespace of the secrets.
+func (b *BootstrapSecretRotator) listBootstrapSecrets() []*corev1.Secret {
+	// informers return slice of pointers so we'll convert them to value types before returning
+	secretPtrs := b.informerCollection.List(informers.InformerKeySecret)
+	var secrets []*corev1.Secret
+
+	for _, secretPtr := range secretPtrs {
+		if secretPtr == nil {
+			continue
+		}
+		secret, ok := secretPtr.(*corev1.Secret)
+		if !ok {
+			continue
+		}
+		// finds bootstrap secrets
+		if strings.Contains(secret.Name, bootstrapSecretPrefix) {
+			secrets = append(secrets, secret)
+		}
+	}
+
+	return secrets
+}
+
+// rotateBootstrapSecrets finds the bootstrap secret of the given certificate
+// from the list of secrets stored in the informerCollection's store and updates the secret.
+func (b *BootstrapSecretRotator) rotateBootstrapSecrets() {
+	secrets := b.listBootstrapSecrets()
+	for _, secret := range secrets {
+		secretProxyUUID := strings.ReplaceAll(secret.Name, bootstrapSecretPrefix, "")
+		certs := b.certManager.ListIssuedCertificates()
+		for _, cert := range certs {
+			// CommonName for a xds certificate has the form: <ProxyUUID>.<kind>.<identity>
+			certProxyUUID := strings.Split(cert.CommonName.String(), ".")[0]
+			// checks if this is the corresponding cert for the secret
+			if secretProxyUUID != certProxyUUID {
+				continue
+			}
+			opts := []certificate.IssueOption{certificate.FullCNProvided()}
+			c, err := b.certManager.IssueCertificate(cert.CommonName.String(), cert.CertType, opts...)
+			if err != nil {
+				log.Error().Err(err).Str(errcode.Kind, errcode.GetErrCodeWithMetric(errcode.ErrRotatingCert)).
+					Msgf("Error rotating cert SerialNumber=%s", cert.GetSerialNumber())
+			}
+			// if the secret and issued cert are the same no need to update the secret
+			if reflect.DeepEqual(c.GetTrustedCAs, secret.Data["ca.crt"]) && reflect.DeepEqual(c.PrivateKey, secret.Data["tls.key"]) && reflect.DeepEqual(c.CertChain, secret.Data["tls.crt"]) {
```

Comment:
It would be good to also check expiration times. This would allow us reduce the expiration on bootstrap certs from 10 years. Not required as of now however.

@jaellio curious on your thoughts

Reply:
Are you suggesting making the bootstrap certificate expiration configurable?

Reply:
we can make it configurable later if we choose, but yes if we're adding rotation functionality we may as well check if the certs are expiring